### PR TITLE
feat(enhancement): Make motherships and carried ships recharge each other from battery power

### PR DESCRIPTION
**Pull request by tibetiroka**
**Feature:** This PR makes fighters and carriers use battery power to recharge each other's batteries.

## Feature Details
This is an enhancement mainly directed at battery powered fighters, but it can be useful in any number of situations. I talked to Sam about it; he doesn't have this in any of his PRs, and gave a thumbs up to the idea.

If the carried ships have deployment orders, their batteries are recharged from the mothership's batteries to cut down the recharge times, and deploy sooner.

If the carried ships were recalled, they replenish the mothership's batteries to help it get away, since the fighters don't need the energy anyway.

If there is excess energy generated by the carrier, that will continue recharge the fighters' batteries regardless of orders, as before.

Currently, the battery power transfer is instantaneous. It might look better in the UI if its progressive, like the old generator recharge; I'm thinking limiting the depletion/fill percentages to, say, 10%/second. I'm looking for opinions on this one.

## Testing Done
Tried it with a battery-only fighter and flagship. Works as expected. Save file: [Test Pilot.txt](https://github.com/endless-sky/endless-sky/files/10754968/Test.Pilot.txt)

### Automated Tests Added
N/A

## Performance Impact
N/A

### CODE REVIEW

File: source/Ship.cpp [hunk 1]
```diff
@@ -2306,6 +2306,16 @@ void Ship::DoGeneration()
 		shieldDelay = max(0, shieldDelay - 1);
 		hullDelay = max(0, hullDelay - 1);
 	}
+	// Carried ships can steal energy from their parent's batteries,
```
Comment (RisingLeaf) [inline source/Ship.cpp:2309]:
```suggestion
	// Carried ships can recharge energy from their parent's batteries,
```

**Review by RisingLeaf — APPROVED** [review]
If you tested this, I think this looks good.
just a small wording issue

File: source/Ship.cpp [hunk 2]
```diff
@@ -2306,6 +2306,16 @@ void Ship::DoGeneration()
...
+	// if they are preparing for deployment. Otherwise, they replenish the
+	// parent's batteries.
+	if(!GetSystem())
```
Comment (quyykk) [inline source/Ship.cpp:2312]: Is this correct? When you're *not* in a system?

Comment (tibetiroka) [reply source/Ship.cpp:2312]: When you are carried by a ship.

Comment (RisingLeaf) [reply source/Ship.cpp:2312]: yea, should be the check for being carried

Comment (quyykk) [reply source/Ship.cpp:2312]: Ooh right

File: source/Ship.cpp [hunk 3]
```diff
@@ -2306,6 +2306,16 @@ void Ship::DoGeneration()
...
+	// Carried ships can recharge energy from their parent's batteries,
+	// if they are preparing for deployment. Otherwise, they replenish the
+	// parent's batteries.
+	if(!GetSystem())
+	{
+		if(HasDeployOrder())
+			DoRepair(energy, GetParent()->energy, attributes.Get("energy capacity"));
+		else
+			DoRepair(GetParent()->energy, energy, GetParent()->Attributes().Get("energy capacity"));
+	}
```
Comment (quyykk) [inline source/Ship.cpp:2318]: Why not put this after the carried ships recharge code above?

File: source/Ship.cpp [hunk 4]
```diff
@@ -2300,6 +2300,18 @@ void Ship::DoGeneration()
...
+			// Carried ships can steal energy from their parent's batteries,
+			// if they are preparing for deployment. Otherwise, they replenish the
+			// parent's batteries.
+			for(const pair<double, Ship *> &it : carried)
+			{
+				Ship &ship = *it.second;
+				if(ship.HasDeployOrder())
+					DoRepair(ship.energy, energy, ship.attributes.Get("energy capacity"));
+				else
+					DoRepair(energy, ship.energy, Attributes().Get("energy capacity"));
```
Comment (quyykk) [inline source/Ship.cpp:2313]:
```suggestion
					DoRepair(energy, ship.energy, attributes.Get("energy capacity"));
```

File: source/Ship.cpp [hunk 5]
```diff
@@ -2300,6 +2300,18 @@ void Ship::DoGeneration()
 				if(fuelRemaining > 0.)
 					DoRepair(ship.fuel, fuelRemaining, ship.attributes.Get("fuel capacity"));
 			}
+
+			// Carried ships can steal energy from their parent's batteries,
```
Comment (TomGoodIdea) [inline source/Ship.cpp:2304]: Why this got reverted?
```suggestion
			// Carried ships can recharge energy from their parent's batteries,
```

Comment (tibetiroka) [reply source/Ship.cpp:2304]: No idea. Probably a mishandled merge conflict or something.

**Review by Hurleveur — APPROVED** [review]
nice! giving some additional use to carried ships, cant wait to make them recharge my solifuge's shields at insane rate

**Comment by Hurleveur** [comment]
Just gets me thinking, shouldnt we update some tooltip or some display to show this extra power potential?

**Comment by tibetiroka** [comment]
I don't recall any tooltips saying there is no energy transfer, and this is a pretty intuitive feature, so I don't think this needs additional explanation.

**Review by quyykk — APPROVED** [review]

**Pull request merged**